File: leetcode/js/convert-an-array-into-a-2d-array-with-conditions.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var findMatrix = function (nums) {
  const count = new Array(nums.length + 1).fill(0);
  const result = [];

  for (const num of nums) {
    if (result.length <= count[num]) {
      result.push([]);
    }

    result[count[num]].push(num);
    count[num]++;
  }

  return result;
};

File: leetcode/rust/convert-an-array-into-a-2d-array-with-conditions.rs
impl Solution {
    pub fn find_matrix(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut count = vec![0; nums.len() + 1];
        let mut result = Vec::new();

        for num in nums {
            let cur_count = count[num as usize];

            if result.len() <= cur_count {
                result.push(Vec::new());
            }

            result[cur_count].push(num);
            count[num as usize] += 1;
        }

        return result;
    }
}
